torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -1655,9 +1655,7 @@ void ProcessGroupNCCL::heartbeatMonitor() {
     // requests, but this is fine since all PG's feed into the same flight
     // recorder and dump. After dump, the training should continue.
     if (dumpPipe.has_value() && dumpPipe->shouldDump()) {
-      // best effort dump, not waiting for the dump here
-      std::future<bool> fut = std::async(
-          std::launch::async, [this]() { return this->dumpDebuggingInfo(); });
+      this->dumpDebuggingInfo();
```

Comment:
interesting, so your change is equivalent?  but we didn't want to block here, so we probably actually want a different change.. we have to make the async object outlive the function?

not sure if making this dump async is critical or not. but it may have been causing problems that we failed to diagnose because of making bad assumptions.

cc @fduwjj @shuqiangzhang

Reply:
This one is for the dump from a pipe which is less used. But yes, this is a good point, let me follow up on this.

Reply:
@wconstab  Even if the future object is placed outside the block, it still blocks when the whole function returns, is that desired?

Reply:
@cyyever we were chatting about stashing Futures we don't want to wait for to a garbage collector at global scope. Something like this:
```
future = std::async(std::launch::async, func_I_dont_wait_for);
futGC.push_back(std::move(future));
```
And then at ProcessGroup, we did a check:
```
if (!futGC.empty()) {
    WARN(“There are unfinished tasks associated with this PG but this PG is destructing. Your program may experience undefined behavior.”);
}
```

Reply:
At what time are finished works in futGC poped so that it is not empty? Before checking empty?

Reply:
@kwen2501 Implements a generic future pool may be a good idea to discover new optimisation opportunities.

Reply:
We have some side threads in ProcessGroup that's constantly doing monitoring (on the collectives). Those threads can pop finished Futures from futGC.

> Implements a generic future pool may be a good idea to discover new optimisation opportunities

Sounds good!

Reply:
Sync offline with @kwen2501 If this is only for FR dump. Why can't we always do this:
1. Always wait for the async for a short period of time.
2. Always throw exceptions when it's timeout or return false. 

This is like what we did `waitForFutureOrTimeout`